tests/integration/index/query_with_relation_filter_test.go

```diff
@@ -553,3 +553,113 @@ func TestQueryWithIndexOnOneToOne_IfFilterOnIndexedRelation_ShouldFilter(t *test
 
 	testUtils.ExecuteTestCase(t, test)
 }
+
+func TestQueryWithIndexOnManyToOne_IfFilterOnIndexedField_ShouldFilterWithExplain(t *testing.T) {
+	req := `query {
+		Device(filter: {
+			year: {_eq: 2021}
+		}) {
+			model
+			owner {
+				name
+			}
+		}
+	}`
+	test := testUtils.TestCase{
+		Description: "With filter on indexed field of secondary relation (N-1) should fetch secondary and primary objects",
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: `
+					type User {
+						name: String
+						devices: [Device]
+					} 
+
+					type Device {
+						model: String 
+						year: Int @index
+						owner: User
+					}
+				`,
+			},
+			testUtils.CreatePredefinedDocs{
+				Docs: getUserDocs(),
+			},
+			testUtils.Request{
+				Request: req,
+				Results: []map[string]any{
+					{
+						"model": "Playstation 5",
+						"owner": map[string]any{
+							"name": "Islam",
+						},
+					},
+					{
+						"model": "Playstation 5",
+						"owner": map[string]any{
+							"name": "Addo",
+						},
+					},
+					{
+						"model": "iPhone 10",
+						"owner": map[string]any{
+							"name": "Addo",
+						},
+					},
+				},
+			},
+			testUtils.Request{
+				Request:  makeExplainQuery(req),
+				Asserter: testUtils.NewExplainAsserter().WithFieldFetches(9).WithIndexFetches(3),
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestQueryWithIndexOnManyToOne_IfFilterOnIndexedRelation_ShouldFilterWithExplain(t *testing.T) {
```

Comment:
todo: This test is really painful to try and read, to try and figure out what has/hasn't been filtered out I am required to read (and hold in memory) nearly 500 lines worth of documents.  Can you please simplify this a little bit so that we can see what is going on.

Reply:
I'm not quite sure what you mean by "simplify". The whole test is around 40 lines of code.
I added some comments.

Reply:
I cannot readily see the relationship between the (hidden) inputs, and the asserted outputs.  At the moment this test is not very valuable as documentation - it only asserts that the current (obscured) behaviour does not change, it does not describe what the current behaviour is (in a way that is readable to humans).

One of the bugs that was fixed last week was leaked due to a similar failing in the tests.  It was tested, however the test asserted the behaviour that was observed, not the behaviour that was desired - and because the test was impossible to read the bug was discovered by a partner instead of us during development.

Reply:
> The whole test is around 40 lines of code.

If you don't care about what data the test is querying, if you do, it is closer to 500 lines.

Reply:
> At the moment this test is not very valuable as documentation

no tests are particularly valuable as documentation if they contain over 100 lines of code with most of them being copy&paste of one another.

> it does not describe what the current behaviour is (in a way that is readable to humans).

Are you sure you are referring to the right test? The test under question initially failed, which means it wasn't an observable behaviour. In order to make it green - so that it behaves in the desired way - I had to change production code.

> One of the bugs that was fixed last week was leaked due to a similar failing in the tests. It was tested, however the test asserted the behaviour that was observed, not the behaviour that was desired - and because the test was impossible to read the bug was discovered by a partner instead of us during development.

I'm not sure if you intentionally phrased it this way oversimplifying. I can give you a dozen of other factor thats contribute to it.

The test are possible to read. I would kindly ask you to write next time "it was had for me to read" so that we don't spend time arguing about subjective matter and try find a solution instead.

Reply:
> Are you sure you are referring to the right test?

This test, `TestQueryWithIndexOnManyToOne_IfFilterOnIndexedRelation_ShouldFilterWithExplain`.  It is hard to tell why the results that are returned, are returned, and why those devices not returned are not returned - because the documents under test are not visible.

It probably seems far more readable to you because you wrote it, and have a stronger memory of the 500 lines of documents under test.  Not everyone, including reviewers, and your future self will have that benefit.

The amount of extra clutter in the test documents also clutters debugging - having a lot of extra stuff (e.g. addresses) floating around in the system really does not help narrowing down the source of any future bugs.

> I can give you a dozen of other factor thats contribute to it.

Sure, and we should work on reducing those too, but that doesn't mean we cant also improve the tests.

> The test are possible to read.

For you, the author, and that's great, but the code needs to be read and maintained by others too.

Reply:
I agree that it makes debugging less trivial. And I understand that it might be a bit annoying to have to jump to another place where all the docs are defined.
We discussed it already in a big round about the different styles that we prefer and there was obvious winner.
I'm willing to stick more to your preferred way, but for the indexes I chose this way because it's not enough to just have the documents that are involved in queries, but also a number of other documents in the db to make sure those are not involved, i.e. to see clear number in `execInfo`.

Anyway I think whatever we want to do should not be part of this PR.

Reply:
The test(s) in *this* PR are not particularly readable.   It don't particularly care which 'style' they are written, so long as they are as easy to read and maintain as they can be.  This means testing with a dataset containing only useful elements, and having both the input and outputs visible in the test.

I'm pushing back hard now, because I feel like we have been far too lax lately, and a lot of expensive developer time has been wasted recently by sub-optimal testing.

Having a personal style is fine, so long as it doesn't waste your colleagues' (and users') time.  We all need to be able to read and maintain the tests, not just you.

Reply:
I told you about the reasons why "only useful elements" is not good enough for testing secondary indexes. You ignored this part.
This means besides "useful elements" there should be nearly the same amount of other not-so-useful document.

The test that you changed recently `TestQueryWithIndexOnOneToMany_IfFilterOnIndexedRelation_ShouldFilterWithExplain` uses only "useful elements" which doesn't really show the power of indexes as it touches all documents in the db the same way full scan does. I didn't have manage to review it before it was merge, otherwise I would point it out.

> The test(s) in this PR are not particularly readable. 

I'd like to get somebody's opinion on this. @fredcarle do you also feel this way? If yes, I will just change. Tired of discussing this.

Reply:
> Do you think addresses and device specs are useful to this test? They are never queried. Most of the fields on the two collections used are not touched either. Anyone reading or debugging the test is forced to deal with them.

holy cow! Now I see why you are concerned so much about the setup. 
But this is not how generation of predefined documents work. All the extra collections and even fields of collections that are not defined in the schema (given to `SchemaUpdate`) are ignored. Please refer the the readme of the package https://github.com/sourcenetwork/defradb/tree/develop/tests/predefined

Reply:
Hahahaha :grin: That does make things an awful lot better than I thought :grin: 

It doesn't really remove the readabilty problem though (and in some ways makes it worse, although the trade off is well worth it in order to trim the data down).

I think documentation on the test action needs to be expanded, that readme is not as useful as it could be if it were linked to from the action, with a brief 2 line summary (out of scope here, not asking for that in this PR :))

That's nice though, we can largely ignore my prior debugging concerns here :)

I think with https://github.com/sourcenetwork/defradb/issues/2592 the importance of the normal non-explain request would be greatly reduced, as the X results are returned would be covered by our normal tests, so perhaps we can focus on explaining the explain tests here.

suggstion: Could you perhaps add a comment explaining the `44` and the `1` explain request expected values? Both here and in the other 2 tests in this PR.  Given the size of `getUserDocs` and the magic of `CreatePredefinedDocs` it remains very hard to understand what those values should be and how they are what they are.

Reply:
Created an issue for the action documentation here: https://github.com/sourcenetwork/defradb/issues/2600

Reply:
added comments to tests and to predefined docs action

Reply:
> added comments to tests and to predefined docs action

They look good, and thanks for sorting out #2600 here too!